staging/utils/signals/handler.go

```diff
@@ -0,0 +1,140 @@
+package signals
+
+import (
+	"context"
+	"errors"
+	"os"
+	"os/signal"
+	"runtime"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/seal-io/walrus/utils/log"
+)
+
+var registered = make(chan struct{})
+
+// Handler registers for signals and returns a context.
+func Handler() context.Context {
+	close(registered) // Panics when called twice.
+
+	logger := log.WithName("signal").WithName("handler")
+
+	sigs := []os.Signal{syscall.SIGINT, syscall.SIGTERM}
+	isRootProc := syscall.Getpid() == 1
+
+	if isRootProc { // Reap child processes.
+		sigs = append(sigs, syscall.SIGCHLD)
+	}
+
+	sigChan := make(chan os.Signal, len(sigs))
+	reapChan := make(chan os.Signal)
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// Register for signals.
+	signal.Notify(sigChan, sigs...)
+
+	// Process signals.
+	go func() {
+		var exited bool
+
+		for sig := range sigChan {
+			logger.V(5).Infof("received signal %q", sig)
+
+			if !exited && sig == syscall.SIGCHLD {
+				select {
+				case reapChan <- sig:
+				default:
+					// Don't block if the channel is full.
+				}
+
+				continue
+			}
+
+			if exited {
+				os.Exit(1)
+			}
+
+			logger.Info("exiting")
+			cancel()
+			exited = true
+
+			close(reapChan)
+		}
+	}()
+
+	if !isRootProc {
+		return ctx
+	}
+
+	// Reap child processes.
+	go func() {
+		for range reapChan {
+			const zombieTimeWait = 10 * time.Millisecond
+
+			for {
+				runtime.Gosched()
+
+				// Use waitid to get an exited child process,
+				// not like waitpid, waitid can leave the child process in waitable state.
+				var si Siginfo
+
+				err := syscallWaitid(P_ALL, 0, &si, syscall.WEXITED|syscall.WNOWAIT, nil)
+				for errors.Is(err, syscall.EINTR) { // Try again if got interrupted.
```

Comment:
What will happen if error is not syscall.EINTR

Reply:
yes